File: problem/F4-Network_test_maxcho/F4_dockerfile+url+ip/src/main.py
#!/usr/bin/env python3
"""
Network Test Client - Tests Docker containers, IP/URL connectivity
Reads test specification from stdin and outputs detailed debug logs
"""
import socket
import sys
import json


def debug_log(msg):
    """Print debug message with prefix"""
    print(f"[DEBUG] {msg}", flush=True)


def check_connection(host, port, timeout=5):
    """Attempt to connect to host:port and return (success, error_msg)"""
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(timeout)

        debug_log(f"Attempting connection to {host}:{port}")

        # Try DNS resolution first if it's a hostname
        if not host.replace('.', '').isdigit():
            try:
                ip = socket.gethostbyname(host)
                debug_log(f"DNS resolved {host} -> {ip}")
                if ip == "0.0.0.0":
                    debug_log("DNS sinkholed to 0.0.0.0")
                    return False, "DNS sinkholed"
            except socket.gaierror as e:
                debug_log(f"DNS resolution failed: {e}")
                return False, f"DNS failed: {e}"

        result = sock.connect_ex((host, port))
        sock.close()

        if result == 0:
            debug_log(f"Connection successful!")
            return True, None
        else:
            debug_log(f"Connection failed with errno={result}")
            return False, f"errno={result}"

    except Exception as e:
        debug_log(f"Connection exception: {e}")
        return False, str(e)


def send_http_request(host,
                      port,
                      path="/",
                      method="GET",
                      body=None,
                      timeout=5):
    """Send HTTP request and return response body"""
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(timeout)

        debug_log(f"Connecting to {host}:{port}")
        sock.connect((host, port))
        debug_log("Connected!")

        # Build request
        if body:
            request = f"{method} {path} HTTP/1.0\r\nHost: {host}\r\nContent-Length: {len(body)}\r\n\r\n{body}"
        else:
            request = f"{method} {path} HTTP/1.0\r\nHost: {host}\r\n\r\n"

        debug_log(f"Sending {method} request to {path}")
        sock.sendall(request.encode())

        # Receive response
        debug_log("Receiving response...")
        response = b""
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            response += chunk

        sock.close()

        response_str = response.decode("utf-8", errors="replace")
        debug_log(f"Response length: {len(response_str)} bytes")

        # Extract body (after double CRLF)
        parts = response_str.split("\r\n\r\n", 1)
        if len(parts) > 1:
            body = parts[1]
            debug_log(f"Body: {body[:200]}...")
            return True, body
        else:
            debug_log(f"No body found in response")
            return True, response_str

    except Exception as e:
        debug_log(f"HTTP request failed: {e}")
        return False, str(e)


def test_docker_env(env_name, port, expected_signature):
    """Test connection to a Docker environment container"""
    debug_log(f"Testing Docker env: {env_name}:{port}")
    debug_log(f"Expected signature: {expected_signature}")

    success, body = send_http_request(env_name, port)

    if success and expected_signature in body:
        debug_log(f"Signature matched!")
        return True
    else:
        debug_log(f"Signature NOT found in response")
        return False


def test_ip_connectivity(ip, port, expect_success=True):
    """Test raw IP connectivity"""
    debug_log(
        f"Testing IP: {ip}:{port}, expect={'connect' if expect_success else 'block'}"
    )

    success, error = check_connection(ip, port)

    if success == expect_success:
        debug_log(
            f"Result as expected: {'connected' if success else 'blocked'}")
        return True
    else:
        debug_log(
            f"Unexpected result: {'connected' if success else 'blocked'}")
        return False


def test_url_connectivity(url, port, expect_success=True):
    """Test URL connectivity (includes DNS resolution)"""
    debug_log(
        f"Testing URL: {url}:{port}, expect={'connect' if expect_success else 'block'}"
    )

    success, error = check_connection(url, port)

    if success == expect_success:
        debug_log(
            f"Result as expected: {'connected' if success else 'blocked'}")
        return True
    else:
        debug_log(
            f"Unexpected result: {'connected' if success else 'blocked'}")
        return False


def parse_test_spec(line):
    """Parse test specification line
    Format: TYPE TARGET PORT [EXPECT] [SIGNATURE]
    Examples:
        DOCKER env-python 8000 "Hello from Server Container!"
        IP 1.1.1.1 443 connect
        URL www.google.com 443 connect
        IP 8.8.8.8 443 block
    """
    parts = line.strip().split()
    if len(parts) < 3:
        return None

    test_type = parts[0].upper()
    target = parts[1]
    port = int(parts[2])

    # Default expectations
    expect = "connect"
    signature = None

    if len(parts) >= 4:
        if test_type == "DOCKER":
            # For DOCKER, remaining parts are the signature
            signature = " ".join(parts[3:])
        else:
            expect = parts[3].lower()

    return {
        "type": test_type,
        "target": target,
        "port": port,
        "expect": expect,
        "signature": signature
    }


def run_test(spec):
    """Run a single test based on specification"""
    test_type = spec["type"]
    target = spec["target"]
    port = spec["port"]
    expect = spec["expect"]
    signature = spec["signature"]

    print(f"\n{'='*60}")
    print(f"Test: {test_type} {target}:{port}")
    print(f"{'='*60}")

    if test_type == "DOCKER":
        result = test_docker_env(target, port, signature if signature else "")
    elif test_type == "IP":
        result = test_ip_connectivity(target, port, expect == "connect")
    elif test_type == "URL":
        result = test_url_connectivity(target, port, expect == "connect")
    else:
        debug_log(f"Unknown test type: {test_type}")
        result = False

    status = "PASS" if result else "FAIL"
    print(f"Result: [{status}]")
    return result


def main():
    print("=" * 60)
    print("Network Test Client (Python)")
    print("=" * 60)

    # Read all input
    input_data = sys.stdin.read().strip()
    if not input_data:
        debug_log("No input provided!")
        return

    lines = input_data.split('\n')
    debug_log(f"Processing {len(lines)} test(s)")

    results = []
    for line in lines:
        line = line.strip()
        if not line or line.startswith('#'):
            continue

        spec = parse_test_spec(line)
        if spec:
            result = run_test(spec)
            results.append(result)
        else:
            debug_log(f"Invalid test spec: {line}")

    print(f"\n{'='*60}")
    print(f"Summary: {sum(results)}/{len(results)} tests passed")
    print(f"{'='*60}")


if __name__ == "__main__":
    main()

File: problem/F4-Network_test_maxcho/F4_dockerfile+url+ip/src/main.c
#include <arpa/inet.h>
#include <errno.h>
#include <netdb.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <unistd.h>

void debug_log(const char *msg) {
  printf("[DEBUG] %s\n", msg);
  fflush(stdout);
}

void debug_logf(const char *fmt, ...) {
  char buffer[512];
  va_list args;
  va_start(args, fmt);
  vsnprintf(buffer, sizeof(buffer), fmt, args);
  va_end(args);
  printf("[DEBUG] %s\n", buffer);
  fflush(stdout);
}

int is_ip_address(const char *str) {
  struct in_addr addr;
  return inet_pton(AF_INET, str, &addr) == 1;
}

int check_connection(const char *host, int port, int timeout_sec) {
  struct sockaddr_in server;
  char log_buf[256];

  server.sin_family = AF_INET;
  server.sin_port = htons(port);

  if (!is_ip_address(host)) {
    snprintf(log_buf, sizeof(log_buf), "Resolving hostname: %s", host);
    debug_log(log_buf);

    struct hostent *he = gethostbyname(host);
    if (!he) {
      debug_log("DNS resolution failed");
      return 0;
    }
    server.sin_addr = *((struct in_addr *)he->h_addr_list[0]);

    char *resolved = inet_ntoa(server.sin_addr);
    snprintf(log_buf, sizeof(log_buf), "Resolved to: %s", resolved);
    debug_log(log_buf);

    if (strcmp(resolved, "0.0.0.0") == 0) {
      debug_log("DNS sinkholed");
      return 0;
    }
  } else {
    inet_pton(AF_INET, host, &server.sin_addr);
  }

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    debug_log("Socket creation failed");
    return 0;
  }

  struct timeval tv;
  tv.tv_sec = timeout_sec;
  tv.tv_usec = 0;
  setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));
  setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

  snprintf(log_buf, sizeof(log_buf), "Connecting to %s:%d", host, port);
  debug_log(log_buf);

  int result = connect(sock, (struct sockaddr *)&server, sizeof(server)) >= 0;
  close(sock);

  if (result) {
    debug_log("Connection successful");
  } else {
    snprintf(log_buf, sizeof(log_buf), "Connection failed: %s",
             strerror(errno));
    debug_log(log_buf);
  }
  return result;
}

int send_http_request(const char *host, int port, const char *path,
                      char *response, int resp_size) {
  struct sockaddr_in server;

  server.sin_family = AF_INET;
  server.sin_port = htons(port);

  if (!is_ip_address(host)) {
    struct hostent *he = gethostbyname(host);
    if (!he)
      return 0;
    server.sin_addr = *((struct in_addr *)he->h_addr_list[0]);
  } else {
    inet_pton(AF_INET, host, &server.sin_addr);
  }

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1)
    return 0;

  struct timeval tv;
  tv.tv_sec = 5;
  tv.tv_usec = 0;
  setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    close(sock);
    return 0;
  }

  char request[512];
  snprintf(request, sizeof(request), "GET %s HTTP/1.0\r\nHost: %s\r\n\r\n",
           path, host);
  send(sock, request, strlen(request), 0);

  int total = 0;
  int len;
  while ((len = recv(sock, response + total, resp_size - total - 1, 0)) > 0) {
    total += len;
    if (total >= resp_size - 1)
      break;
  }
  response[total] = '\0';
  close(sock);

  char log_buf[64];
  snprintf(log_buf, sizeof(log_buf), "Response length: %d bytes", total);
  debug_log(log_buf);

  return 1;
}

int test_docker_env(const char *env_name, int port, const char *signature) {
  char log_buf[256];
  snprintf(log_buf, sizeof(log_buf), "Testing Docker env: %s:%d", env_name,
           port);
  debug_log(log_buf);
  snprintf(log_buf, sizeof(log_buf), "Expected signature: %s", signature);
  debug_log(log_buf);

  char response[4096];
  if (!send_http_request(env_name, port, "/", response, sizeof(response))) {
    debug_log("HTTP request failed");
    return 0;
  }

  if (strstr(response, signature)) {
    debug_log("Signature matched!");
    return 1;
  } else {
    debug_log("Signature NOT found");
    return 0;
  }
}

int test_connectivity(const char *target, int port, int expect_connect) {
  char log_buf[128];
  snprintf(log_buf, sizeof(log_buf), "Testing: %s:%d, expecting %s", target,
           port, expect_connect ? "connect" : "block");
  debug_log(log_buf);

  int result = check_connection(target, port, 5);

  if (result == expect_connect) {
    debug_log("Result as expected");
    return 1;
  } else {
    debug_log("Unexpected result");
    return 0;
  }
}

int main() {
  printf("============================================================\n");
  printf("Network Test Client (C)\n");
  printf("============================================================\n");

  char line[512];
  int total = 0, passed = 0;

  while (fgets(line, sizeof(line), stdin)) {
    // Remove newline
    line[strcspn(line, "\r\n")] = 0;

    if (strlen(line) == 0 || line[0] == '#')
      continue;

    char type[32], target[128], expect_str[32], signature[256];
    int port;

    int parts =
        sscanf(line, "%31s %127s %d %255[^\n]", type, target, &port, signature);
    if (parts < 3)
      continue;

    printf("\n------------------------------------------------------------\n");
    printf("Test: %s %s:%d\n", type, target, port);
    printf("------------------------------------------------------------\n");

    int result = 0;

    if (strcmp(type, "DOCKER") == 0) {
      result = test_docker_env(target, port, signature);
    } else if (strcmp(type, "IP") == 0 || strcmp(type, "URL") == 0) {
      int expect_connect = 1;
      if (parts >= 4 && strcmp(signature, "block") == 0) {
        expect_connect = 0;
      }
      result = test_connectivity(target, port, expect_connect);
    } else {
      char log_buf[64];
      snprintf(log_buf, sizeof(log_buf), "Unknown test type: %s", type);
      debug_log(log_buf);
    }

    printf("Result: [%s]\n", result ? "PASS" : "FAIL");
    total++;
    if (result)
      passed++;
  }

  printf("\n============================================================\n");
  printf("Summary: %d/%d tests passed\n", passed, total);
  printf("============================================================\n");

  return 0;
}